spec/src/main/asciidoc/circuitbreaker.asciidoc

```diff
@@ -59,48 +64,53 @@ public Connection serviceA() {
    return conn;
 }
 ----
-Below are a couple scenarios to demonstrate when the circuit opens for the above code snippet.
 
 * Scenario 1
 ** Request 1 - success
 ** Request 2 - failure
-** Request 3 - failure
+** Request 3 - success
 ** Request 4 - success
-** Request 5 - CircuitBreakerOpenException
+** Request 5 - failure
+** Request 6 - CircuitBreakerOpenException 
 
-The Request 4 will trigger the circuit to open. The Request 5 will hit the exception.
+In this scenario, request 5 will trigger the circuit to open because out of the last four requests (the `requestVolumeThreshold`), two failed
+which reaches the `failureRatio` of `0.5`. Request 6 will therefore hit the `CircuitBreakerOpenException`.
 
 * Scenario 2
 ** Request 1 - success
 ** Request 2 - failure
-** Request 3 - success
+** Request 3 - failure
 ** Request 4 - success
-** Request 5 - failure
-** Request 6 - CircuitBreakerOpenException 
+** Request 5 - CircuitBreakerOpenException
 
-The Request 5 will trigger the circuit to open. The Request 6 will hit the exception.
+In this scenario, request 4 will cause the circuit to open. Request 5 will hit the `CircuitBreakerOpenException`.
+Note that request 3 does not cause the circuit to open because the rolling window has not yet reached the `requestVolumeThreshold`.
 
-=== Circuit Breaker Usage
+==== Configuring which exceptions are considered a failure
 
-A method or a class can be annotated with `@CircuitBreaker`, which means the method or the methods under the class will have CircuitBreaker policy applied.
+The `failOn` and `skipOn` parameters are used to define which exceptions are considered failures for the purpose of deciding whether the circuit breaker should open.
+
+When a method returns a result, the following rules are applied to determine whether the result is a success or a failure:
+
+* If the method does not throw a `Throwable`, it is considered a success
+* Otherwise, if the thrown object is assignable to any value in the `skipOn` parameter, is is considered a success
+* Otherwise, if the thrown object is assignable to any value in the `failOn` parameter, it is considered a failure
+* Otherwise it is considered a success
+
+If a method throws a `Throwable` which is not a subclass of either `Error` or `Exception`, non-portable behavior results.
+
+In the following example, all exceptions assignable to `ExceptionA` and `ExceptionB`, except the ones assignable to `ExeptionBSub` will be considered failures. `ExceptionBSub` and all other exceptions will be considered successes.
```

Comment:
I would not add the sentence`ExceptionBSub` and all other exceptions will be considered successes.`, because they don't contribute towards circuitbreaker not they are not treated as success.

Reply:
Hmm, I put that because, as I understand it, the method throwing an exception which is not considered a failure should be treated the same as the method returning without throwing an exception.

I.e.:
* In closed state, a success is recorded in the rolling window
* In half-open state, the trial execution is considered to have "succeeded" and will allow the circuit breaker to close

I thought this behaviour might be considered unintuitive (why should an exception be considered a success) so I thought it would be helpful to call it out explicitly.

Reply:
If that's not the case and exceptions matching `skipOn` should be ignored completely then I would want to revise the language in the spec and check that we have TCKs to assert that.

Currently the TCK only asserts that an exception matching `skipOn` doesn't cause the breaker to open, but it doesn't check whether it's recorded as a success in the rolling window or whether it will cause the breaker to close when it's in half-open state.

Reply:
`SkipOn` does not change the method exception. It only means the exception is not counted towards the circuit breaker. Similarly, for `Retry`, the `abortOn` means the exception will not trigger `Retry`. The fact of the exception listed in `abortOn` should not cause the Exception to be swallowed. +1 on adding more tests to mandate this.

Reply:
Existing tests do assert that an exception listed in `skipOn` is not swallowed.

I'm still not 100% clear on the desired behaviour when an exception listed in `skipOn` is thrown in terms of:
 * what is stored in the rolling window
 * what should happen if a trial execution ends with an exception listed in `skipOn` being thrown

My assumption was that all of the following cases would be treated as a "success" by the rolling window and for a trial execution:
 * method returns normally
 * method throws exception not listed in `failOn`
 * method throws exception listed in `skipOn`

Only in one case is the execution considered a "failure" by the rolling window and for a trial execution:
 * method throws exception listed in `failOn` and not listed in `skipOn`

I'll continue with this assumption unless someone wants to argue it.

I'll try to update the text to avoid giving the impression that an exception is ever swallowed by the circuit breaker.